Zinc/Private/ZincRepo.m

```diff
@@ -1044,6 +1043,192 @@ - (ZincTask*) queueGarbageCollectTask
+                                                                                      error:&error];
+        totalSize += [fileDictionary fileSize];
+    }
+
+    return totalSize;
+}
+
+- (float)megabytesForBytes:(unsigned long long)bytes {
+    return bytes / 1024.f / 1024.f;
+}
+
+- (unsigned long long)totalSizeInBytesOfBundlesWithPrefix:(NSString *)prefix {
+    __block unsigned long long size = 0;
+
+    [self forBundlesWithPrefix:prefix performBlock:^(NSURL *bundleURL) {
+        size += [self sizeOfFolderInBytesWithPath:bundleURL.path];
+    }];
+
+    return size;
+}
+
+- (float)totalSizeInMegabytesOfBundlesWithPrefix:(NSString *)prefix {
+    unsigned long long bytes = [self totalSizeInBytesOfBundlesWithPrefix:prefix];
+    return [self megabytesForBytes:bytes];
+}
+
+- (NSString *)bundleNameForBundleURL:(NSURL *)bundleURL {
+    return [[bundleURL absoluteString] lastPathComponent];
+}
+
+// Example bundle name: @"com.hello.world-foo-bar-1"
+// Example bundle id: @"com.hello.world-foo-bar"
+- (NSString *)bundleIDForBundleName:(NSString *)bundleName {
+    NSRange rangeOfLastHyphen = [bundleName rangeOfString:@"-"
+                                                  options:NSBackwardsSearch];
+    NSString *bundleID = [bundleName substringWithRange:NSMakeRange(0, rangeOfLastHyphen.location)];
+    return bundleID;
+}
+
+- (NSString *)manifestNameForBundleName:(NSString *)bundleName {
+    return [bundleName stringByAppendingString:@".json"];
+}
+
+- (NSURL *)manifestsFolderURL {
+    return [self absoluteURLForPathRelativeToZincFolder:MANIFESTS_DIR
+                                            isDirectory:YES];
+}
+
+- (NSURL *)manifestURLForManifestName:(NSString *)manifestName {
+    return [[self manifestsFolderURL] URLByAppendingPathComponent:manifestName
+                                                      isDirectory:NO];
+}
+
+- (BOOL)deleteFileAtURL:(NSURL *)url {
+    NSError *error;
+    [NSFileManager.defaultManager removeItemAtURL:url error:&error];
+    if (error) {
+        NSLog(@"Error while deleting file at url: %@ \nfor file path: %@",
+              error.localizedDescription,
+              url.path);
+    }
+
+    return !error;
+}
+
+- (void)purgeBundlesWithPrefix:(NSString *)prefix {
+    NSMutableSet<NSURL *> *bundleURLs = [NSMutableSet new];
+    [self forBundlesWithPrefix:prefix performBlock:^(NSURL *bundleURL) {
+        [bundleURLs addObject:bundleURL];
+    }];
+
+    @synchronized(self.index) {
+        for (NSURL *bundleURL in bundleURLs) {
+            NSString *bundleName = [self bundleNameForBundleURL:bundleURL];
+            NSString *bundleID = [self bundleIDForBundleName:bundleName];
+
+            [self postNotification:ZincRepoBundleWillDeleteNotification
+                          bundleID:bundleID];
+            [self.index purgeBundleWithID:bundleID];
+        }
+    }
+
+    ZincTaskRef* taskRef = [[ZincTaskRef alloc] init];
```

Comment:
In the above use of the lock, the line that is being protected is this
```
 [self.index purgeBundleWithID:bundleID];
```
This does not change any files on disk. This simply changes the `@property`'s of the instance `self.index` (instance of `ZincRepoIndex`)
The actual save to disk is done by
```
[self queueIndexSaveTask]
```

I believe the intention of this pattern is that it doesn't matter if the `self.index` instance has changed between now and when `indexSaveTask` actually happens - it just assumes `self.repo` is always in a valid state. This is a proper assumption, so long as everyone who modifies `self.repo` acquires the `@synchronize(self.repo)` lock when modifying it (such as how we're doing it above)